File: backend/weekaboo/integrations/microsoft_body.py
"""Keep Microsoft's generated meeting HTML separate from editable event notes.

Ranges refer to the original source: protected blocks are copied byte-for-byte,
not parsed and reserialized. The latest Graph body is used on every write.
"""

import re
from dataclasses import dataclass, field
from html import unescape
from html.parser import HTMLParser
from urllib.parse import urlsplit

from .providers.base import ProviderError

VOID = {
    "area",
    "base",
    "br",
    "col",
    "embed",
    "hr",
    "img",
    "input",
    "link",
    "meta",
    "param",
    "source",
    "track",
    "wbr",
}


@dataclass
class Node:
    tag: str
    attrs: dict
    start: int
    content_start: int
    end: int = 0
    close_start: int = 0
    parent: "Node | None" = None
    children: list["Node"] = field(default_factory=list)


class BodyParser(HTMLParser):
    def __init__(self, source: str):
        super().__init__(convert_charrefs=False)
        self.source = source
        self.lines = [0]
        for line in source.split("\n"):
            self.lines.append(self.lines[-1] + len(line) + 1)
        self.nodes: list[Node] = []
        self.stack: list[Node] = []
        self.feed(source)
        self.close()

    def source_offset(self):
        line, col = self.getpos()
        return self.lines[line - 1] + col

    def handle_starttag(self, tag, attrs):
        start = self.source_offset()
        node = Node(
            tag,
            dict(attrs),
            start,
            start + len(self.get_starttag_text()),
            parent=self.stack[-1] if self.stack else None,
        )
        if node.parent:
            node.parent.children.append(node)
        self.nodes.append(node)
        if tag in VOID:
            node.end = node.close_start = node.content_start
        else:
            self.stack.append(node)

    def handle_startendtag(self, tag, attrs):
        self.handle_starttag(tag, attrs)
        if tag not in VOID:
            node = self.stack.pop()
            node.end = node.close_start = node.content_start

    def handle_endtag(self, tag):
        for index in range(len(self.stack) - 1, -1, -1):
            if self.stack[index].tag == tag:
                node = self.stack[index]
                node.close_start = self.source_offset()
                node.end = self.source.find(">", node.close_start) + 1
                del self.stack[index:]
                break

    def protected(self):
        candidates = []
        for node in self.nodes:
            classes = {
                c.removeprefix("x_") for c in (node.attrs.get("class") or "").split()
            }
            identity = (node.attrs.get("id") or "").lower().removeprefix("x_")
            if "me-email-text" in classes or identity in {
                "skypemeeting",
                "teamsmeeting",
                "skypemeetingcontent",
            }:
                if not node.end:
                    raise ProviderError(
                        422,
                        "The meeting details are incomplete. Notes have not been sent; sync and try again.",
                    )
                # Only outermost marked blocks; nested markers must not duplicate.
                if not any(
                    parent.start <= node.start < parent.end for parent in candidates
                ):
                    candidates.append(node)
        # Outlook puts decorative separator divs beside its generated block.
        # Keep those with the block rather than making them editable notes.
        for node in candidates:
            siblings = (
                node.parent.children
                if node.parent
                else [n for n in self.nodes if n.parent is None]
            )
            index = siblings.index(node)
            for neighbor in reversed(siblings[:index]):
                if self.source[neighbor.end : node.start].strip() or not self.separator(
                    neighbor
                ):
                    break
                node.start = neighbor.start
            for neighbor in siblings[index + 1 :]:
                if self.source[node.end : neighbor.start].strip() or not self.separator(
                    neighbor
                ):
                    break
                node.end = neighbor.end
        # Adjacent blocks can both absorb the same separator. Make the ranges
        # disjoint so repeated writes cannot multiply that markup.
        for previous, current in zip(candidates, candidates[1:]):
            current.start = max(current.start, previous.end)
        return candidates

    def separator(self, node):
        if not node.end or node.tag not in {"div", "p", "br", "hr"}:
            return False
        raw = self.source[node.start : node.end]
        if re.search(r"<(?:a|img|input)\b", raw, re.I):
            return False
        text = unescape(re.sub(r"<[^>]*>", "", raw)).strip()
        return not text or bool(re.fullmatch(r"[_—–\-\s]+", text))

    def join_link(self):
        for node in self.nodes:
            if node.tag != "a":
                continue
            try:
                url = urlsplit(node.attrs.get("href") or "")
                if url.hostname in {
                    "teams.live.com",
                    "teams.microsoft.com",
                    "join.skype.com",
                } and (
                    "/meet/" in url.path
                    or "/meetup-join/" in url.path
                    or url.hostname == "join.skype.com"
                ):
                    if (
                        url.scheme in {"http", "https"}
                        and not url.username
                        and not url.password
                    ):
                        return node.attrs["href"]
            except ValueError:
                continue
        return None

    def body_range(self):
        body = next((n for n in self.nodes if n.tag == "body" and n.end), None)
        return (body.content_start, body.close_start) if body else (0, len(self.source))


def editable_notes(source: str | None) -> str | None:
    if not source:
        return source
    parsed = BodyParser(source)
    try:
        blocks = parsed.protected()
    except ProviderError:
        # A malformed provider body must not break calendar reads. Writes still
        # fail closed when merging into that same incomplete body.
        return source
    if not blocks:
        return source
    start, end = parsed.body_range()
    pieces = []
    for block in blocks:
        pieces.append(source[start : block.start])
        start = block.end
    pieces.append(source[start:end])
    return "".join(pieces).strip()


def merge_notes(latest_body: str, notes: str | None, *, online: bool) -> str:
    parsed = BodyParser(latest_body)
    blocks = parsed.protected()
    if online and not blocks and parsed.join_link():
        # Unknown templates must never be guessed at and overwrite joining data.
        raise ProviderError(
            422,
            "This meeting uses an unrecognized details format. Your draft is still here; the remote meeting has not been changed.",
        )
    clean = editable_notes(notes) or ""
    if not blocks:
        return clean
    start, end = parsed.body_range()
    protected_html = "".join(latest_body[node.start : node.end] for node in blocks)
    return latest_body[:start] + clean + protected_html + latest_body[end:]


def meeting_link(source: str | None) -> str | None:
    """Keep a join action for invitations whose only call link is in the body."""
    return BodyParser(source).join_link() if source else None

File: src/engine/providers/microsoft-body.ts
import { parse, type DefaultTreeAdapterTypes } from 'parse5';
import { ServiceError } from '../../services/errors';
type Node = DefaultTreeAdapterTypes.Node;
type Element = DefaultTreeAdapterTypes.Element;
const element = (node: Node): node is Element => 'tagName' in node;
const attribute = (node: Element, name: string) => node.attrs.find(attr => attr.name === name)?.value || '';
const children = (node: Node): Node[] => 'childNodes' in node ? node.childNodes : [];
const text = (node: Node): string => node.nodeName === '#text' ? (node as DefaultTreeAdapterTypes.TextNode).value : children(node).map(text).join('');
function inspect(source: string) {
  const root = parse(source, { sourceCodeLocationInfo: true });
  const nodes: Element[] = [];
  function visit(node: Node) { if (element(node)) nodes.push(node); children(node).forEach(visit); }
  visit(root);
  const marked = (node: Element) => attribute(node, 'class').split(/\s+/).some(name => name.replace(/^x_/, '') === 'me-email-text') || ['skypemeeting', 'teamsmeeting', 'skypemeetingcontent'].includes(attribute(node, 'id').toLowerCase().replace(/^x_/, ''));
  const separator = (node: Node) => element(node) && ['div', 'p', 'br', 'hr'].includes(node.tagName) && !nodes.some(other => ['a', 'img', 'input'].includes(other.tagName) && contains(node, other)) && /^[_—–\-\s]*$/.test(text(node));
  const contains = (outer: Element, inner: Element) => { for (let parent = inner.parentNode; parent; parent = 'parentNode' in parent ? parent.parentNode : null) if (parent === outer) return true; return false; };
  const ranges: { start: number; end: number }[] = [];
  for (const node of nodes.filter(marked)) {
    if (nodes.some(parent => parent !== node && marked(parent) && contains(parent, node))) continue;
    const location = node.sourceCodeLocation;
    // parse5 repairs malformed HTML; never turn a repaired generated block into
    // permission to overwrite Teams details. Require its actual closing tag.
    if (!location?.startTag || !location.endTag) throw new ServiceError('unsupported', 'The meeting details are incomplete. Notes have not been sent; sync and try again.');
    let start = location.startOffset, end = location.endOffset;
    const siblings = node.parentNode ? children(node.parentNode) : [];
    let index = siblings.indexOf(node) - 1;
    while (index >= 0) {
      const sibling = siblings[index--], loc = sibling.sourceCodeLocation;
      if (sibling.nodeName === '#text' && !text(sibling).trim()) continue;
      if (!loc || source.slice(loc.endOffset, start).trim() || !separator(sibling)) break;
      start = loc.startOffset;
    }
    index = siblings.indexOf(node) + 1;
    while (index < siblings.length) {
      const sibling = siblings[index++], loc = sibling.sourceCodeLocation;
      if (sibling.nodeName === '#text' && !text(sibling).trim()) continue;
      if (!loc || source.slice(end, loc.startOffset).trim() || !separator(sibling)) break;
      end = loc.endOffset;
    }
    ranges.push({ start, end });
  }
  ranges.sort((a, b) => a.start - b.start);
  for (let i = 1; i < ranges.length; i++) ranges[i].start = Math.max(ranges[i].start, ranges[i - 1].end);
  const body = nodes.find(node => node.tagName === 'body')?.sourceCodeLocation;
  const start = body?.startTag?.endOffset ?? 0, end = body?.endTag?.startOffset ?? source.length;
  let joinLink: string | undefined;
  for (const node of nodes.filter(node => node.tagName === 'a')) {
    try {
      const url = new URL(attribute(node, 'href'));
      if (['http:', 'https:'].includes(url.protocol) && !url.username && !url.password && ['teams.live.com', 'teams.microsoft.com', 'join.skype.com'].includes(url.hostname) && (url.pathname.includes('/meet/') || url.pathname.includes('/meetup-join/') || url.hostname === 'join.skype.com')) { joinLink = url.href; break; }
    } catch { /* Not a usable absolute meeting link. */ }
  }
  return { ranges, start, end, joinLink };
}
function editable(source: string, parsed: ReturnType<typeof inspect>) {
  if (!parsed.ranges.length) return source;
  let cursor = parsed.start, result = '';
  for (const range of parsed.ranges) { result += source.slice(cursor, range.start); cursor = range.end; }
  return (result + source.slice(cursor, parsed.end)).trim();
}
export function microsoftNotes(source: string) {
  try { const parsed = inspect(source); return { editable: editable(source, parsed), meetingUrl: parsed.joinLink }; }
  catch { return { editable: source, meetingUrl: undefined }; }
}
export function mergeMicrosoftNotes(source: string, notes: string | null, online: boolean) {
  const parsed = inspect(source);
  if (online && !parsed.ranges.length && parsed.joinLink) throw new ServiceError('unsupported', 'This meeting uses an unrecognized details format. Your draft is still here; the remote meeting has not been changed.');
  const cleaned = notes ? editable(notes, inspect(notes)) : '';
  if (!parsed.ranges.length) return cleaned;
  return source.slice(0, parsed.start) + cleaned + parsed.ranges.map(range => source.slice(range.start, range.end)).join('') + source.slice(parsed.end);
}
